File: backend/src/main/java/com/bookstore/service/AuthService.java
package com.bookstore.service;

import com.bookstore.dto.request.LoginRequest;
import com.bookstore.dto.request.RegisterRequest;
import com.bookstore.dto.response.LoginResponse;
import com.bookstore.dto.response.TokenResponse;
import com.bookstore.dto.response.UserResponse;
import com.bookstore.entity.Customer;
import com.bookstore.entity.RefreshToken;
import com.bookstore.entity.Role;
import com.bookstore.entity.SessionEntity;
import com.bookstore.entity.User;
import com.bookstore.event.CustomerGraphProjectionAction;
import com.bookstore.event.CustomerGraphProjectionEvent;
import com.bookstore.repository.CustomerRepository;
import com.bookstore.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.context.ApplicationEventPublisher;
import org.springframework.security.authentication.BadCredentialsException;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.UUID;

@Slf4j
@Service
@RequiredArgsConstructor
public class AuthService {

    private final UserRepository userRepository;
    private final CustomerRepository customerRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtService jwtService;
    private final RefreshTokenService refreshTokenService;
    private final SessionService sessionService;
    private final UserService userService;
    private final ApplicationEventPublisher eventPublisher;

    @Transactional
    public UserResponse register(RegisterRequest request) {
        log.debug("Starting registration process for email: {}", request.getEmail());

        if (userRepository.findByEmail(request.getEmail()).isPresent()) {
            log.warn("Registration failed: Email already exists: {}", request.getEmail());
            throw new IllegalArgumentException("Email already registered");
        }

        User user = User.builder()
                .email(request.getEmail())
                .password(passwordEncoder.encode(request.getPassword()))
                .role(Role.CUSTOMER)
                .isActive(true)
                .build();

        User savedUser = userRepository.save(user);
        log.debug("User created - userId: {}, email: {}", savedUser.getId(), savedUser.getEmail());

        // Create customer profile record with optional fields
        boolean hasFullName = request.getFullName() != null && !request.getFullName().isBlank();
        boolean hasPhone = request.getPhoneNumber() != null && !request.getPhoneNumber().isBlank();
        boolean hasAddress = request.getAddress() != null && !request.getAddress().isBlank();
        boolean profileCompleted = hasFullName && hasPhone && hasAddress;

        Customer customer = Customer.builder()
                .user(savedUser)
                .email(savedUser.getEmail())
                .fullName(hasFullName ? request.getFullName() : null)
                .phoneNumber(hasPhone ? request.getPhoneNumber() : null)
                .address(hasAddress ? request.getAddress() : null)
                .profileCompleted(profileCompleted)
                .build();

        Customer savedCustomer = customerRepository.save(customer);
        log.debug("Customer profile created - customerId: {}, userId: {}, profileCompleted: {}",
                savedCustomer.getId(), savedUser.getId(), profileCompleted);

        publishCustomerGraphProjection(savedUser, CustomerGraphProjectionAction.UPSERT);
        log.info("User registered successfully - userId: {}, customerId: {}, email: {}",
                savedUser.getId(), savedCustomer.getId(), savedUser.getEmail());

        return buildUserResponse(savedUser, savedCustomer);
    }

    @Transactional
    public LoginResponse login(LoginRequest request) {
        log.debug("Attempting login for email: {}", request.getEmail());

        User user = userRepository.findByEmail(request.getEmail())
                .orElseThrow(() -> {
                    log.warn("Login failed: User not found - email: {}", request.getEmail());
                    return new BadCredentialsException("Invalid email or password");
                });

        if (!passwordEncoder.matches(request.getPassword(), user.getPassword())) {
            log.warn("Login failed: Invalid password - email: {}, userId: {}", request.getEmail(), user.getId());
            throw new BadCredentialsException("Invalid email or password");
        }

        if (!user.getIsActive()) {
            log.warn("Login failed: Account deactivated - email: {}, userId: {}", request.getEmail(), user.getId());
            throw new BadCredentialsException("Account has been deactivated. Contact admin.");
        }

        log.debug("Generating tokens for user: {}", user.getId());

        // Create session in Cassandra first
        SessionEntity session = null;
        try {
            String deviceInfo = request.getDeviceInfo() != null ? request.getDeviceInfo() : "Unknown Device";
            String ipAddress = request.getIpAddress() != null ? request.getIpAddress() : "Unknown IP";
            session = sessionService.createSession(user, request.isRememberMe(), deviceInfo, ipAddress);
            log.debug("Session created in Cassandra for userId: {}", user.getId());
        } catch (Exception e) {
            log.error("Failed to create Cassandra session for userId: {}", user.getId(), e);
            // Don't fail login if Cassandra write fails
        }

        // Generate tokens with sessionId
        String accessToken = session != null
            ? jwtService.generateAccessToken(user, session.getSessionId())
            : jwtService.generateAccessToken(user);
        RefreshToken refreshToken = refreshTokenService.createRefreshToken(user, request.isRememberMe());

        log.info("Login successful - userId: {}, email: {}, rememberMe: {}", user.getId(), user.getEmail(), request.isRememberMe());

        // Load customer profile if user is a customer
        Customer customer = null;
        if (user.getRole() == Role.CUSTOMER) {
            customer = customerRepository.findByUserId(user.getId()).orElse(null);
        }

        UserResponse userResponse = buildUserResponse(user, customer);

        return LoginResponse.builder()
                .accessToken(accessToken)
                .refreshToken(refreshToken.getToken())
                .user(userResponse)
                .build();
    }

    public TokenResponse refreshAccessToken(String refreshTokenString) {
        log.debug("Refreshing access token");
        RefreshToken refreshToken = refreshTokenService.validateRefreshToken(refreshTokenString);
        User user = refreshToken.getUser();

        String accessToken = jwtService.generateAccessToken(user);
        log.info("Access token refreshed successfully for userId: {}", user.getId());

        return TokenResponse.builder()
                .accessToken(accessToken)
                .build();
    }

    @Transactional
    public void logout(String refreshToken) {
        if (refreshToken != null) {
            log.debug("Processing logout with refresh token");
            refreshTokenService.deleteByToken(refreshToken);
            log.info("Logout successful - refresh token deleted");
        } else {
            log.debug("Logout called without refresh token");
        }
    }

    private void publishCustomerGraphProjection(User user, CustomerGraphProjectionAction action) {
        eventPublisher.publishEvent(new CustomerGraphProjectionEvent(
                user.getId(),
                user.getEmail(),
                user.getEmail(),
                user.getRole(),
                user.getIsActive(),
                action
        ));
    }

    public UserResponse getCurrentUser(UUID userId) {
        log.debug("Fetching current user for userId: {}", userId);
        User user = userService.findById(userId);
        log.debug("Current user fetched successfully: {}", user.getEmail());

        // Load customer profile if user is a customer
        Customer customer = null;
        if (user.getRole() == Role.CUSTOMER) {
            customer = customerRepository.findByUserId(user.getId()).orElse(null);
        }

        return buildUserResponse(user, customer);
    }

    private UserResponse buildUserResponse(User user, Customer customer) {
        UserResponse.UserResponseBuilder builder = UserResponse.builder()
                .id(user.getId())
                .email(user.getEmail())
                .role(user.getRole())
                .isActive(user.getIsActive())
                .createdAt(user.getCreatedAt());

        // Add customer profile fields if available
        if (customer != null) {
            builder.fullName(customer.getFullName())
                    .phoneNumber(customer.getPhoneNumber())
                    .address(customer.getAddress())
                    .profileCompleted(customer.isProfileCompleted());
        } else {
            builder.fullName(null)
                    .phoneNumber(null)
                    .address(null)
                    .profileCompleted(false);
        }

        return builder.build();
    }
}

File: frontend/src/services/authService.ts
import axiosInstance from '../utils/axiosConfig';
import type { LoginResponse, TokenResponse, User, UserResponse } from '../types';

// Mock data for development (until backend is ready)
const MOCK_MODE = false;

const mockUsers = [
  {
    id: '1',
    email: import.meta.env.VITE_MOCK_ADMIN_EMAIL || 'admin@example.test',
    password: import.meta.env.VITE_MOCK_ADMIN_PASSWORD || 'change-me-admin-password',
    fullName: 'Demo Administrator',
    phoneNumber: '+10000000000',
    address: 'Demo Address',
    role: 'ADMIN' as const,
    isActive: true,
    profileCompleted: true,
    createdAt: new Date().toISOString(),
  },
  {
    id: '2',
    email: 'customer@example.com',
    password: import.meta.env.VITE_MOCK_CUSTOMER_PASSWORD || 'change-me-customer-password',
    fullName: 'Demo Customer',
    phoneNumber: '+10000000001',
    address: 'Demo Address',
    role: 'CUSTOMER' as const,
    isActive: true,
    profileCompleted: true,
    createdAt: new Date().toISOString(),
  },
];

// Simulate API delay
const delay = (ms: number) => new Promise(resolve => setTimeout(resolve, ms));

export const authService = {
  async register(data: {
    email: string;
    password: string;
    fullName: string;
    phoneNumber: string;
    address: string;
  }): Promise<UserResponse> {
    if (MOCK_MODE) {
      await delay(1000);

      // Check if email already exists
      if (mockUsers.find(u => u.email === data.email)) {
        throw {
          response: {
            status: 409,
            data: {
              message: 'Email already exists',
              errors: { email: 'This email is already registered' }
            }
          }
        };
      }

      const newUser = {
        id: String(mockUsers.length + 1),
        ...data,
        role: 'CUSTOMER' as const,
        isActive: true,
        profileCompleted: true,
        createdAt: new Date().toISOString(),
      };
      mockUsers.push(newUser);

      const { password, ...userResponse } = newUser;
      return userResponse;
    }

    const response = await axiosInstance.post<UserResponse>('/auth/register', data);
    return response.data;
  },

  async login(email: string, password: string, rememberMe: boolean): Promise<LoginResponse> {
    if (MOCK_MODE) {
      await delay(1000);

      const user = mockUsers.find(u => u.email === email);

      if (!user || user.password !== password) {
        throw {
          response: {
            status: 401,
            data: {
              message: 'Invalid email or password'
            }
          }
        };
      }

      if (!user.isActive) {
        throw {
          response: {
            status: 403,
            data: {
              message: 'Account has been deactivated. Contact admin.'
            }
          }
        };
      }

      const { password: _, ...userResponse } = user;
      return {
        accessToken: 'mock-access-token-' + user.id,
        user: userResponse,
      };
    }

    // Get device info and IP address
    const deviceInfo = navigator.userAgent;
    const ipAddress = 'client'; // IP will be captured by backend from request

    const response = await axiosInstance.post<LoginResponse>('/auth/login', {
      email,
      password,
      rememberMe,
      deviceInfo,
      ipAddress,
    });
    return response.data;
  },

  async logout(): Promise<void> {
    if (MOCK_MODE) {
      await delay(500);
      return;
    }

    await axiosInstance.post('/auth/logout');
  },

  async refreshToken(): Promise<TokenResponse> {
    if (MOCK_MODE) {
      await delay(500);
      return {
        accessToken: 'mock-refreshed-token',
      };
    }

    const response = await axiosInstance.post<TokenResponse>('/auth/refresh');
    return response.data;
  },

  async getCurrentUser(): Promise<User> {
    if (MOCK_MODE) {
      await delay(500);
      // Return mock admin user
      const { password, ...user } = mockUsers[0];
      return user;
    }

    const response = await axiosInstance.get<User>('/auth/me');
    return response.data;
  },
};
